File: python/sorting/merge_sort.py
"""
Merge Sort
"""


def _merge(array, middle, left, right):
    left_len = middle - left+1
    right_len = right - middle

    left_arr = [0]*left_len
    right_arr = [0]*right_len

    for g in range(0, left_len):
        left_arr[g] = array[left+g]
    for g in range(0, right_len):
        right_arr[g] = array[middle+1+g]

    i = 0
    j = 0
    k = left

    while i < left_len and j < right_len:
        if left_arr[i] <= right_arr[j]:
            array[k] = left_arr[i]
            i += 1
        else:
            array[k] = right_arr[j]
            j += 1
        k += 1
    while i < left_len:
        array[k] = left_arr[i]
        i += 1
        k += 1
    while j < right_len:
        array[k] = right_arr[j]
        j += 1
        k += 1


def merge_sort(array: list, l, r):
    """
    Merge Sort
    :param array: given array
    :param l: beginning
    :param r: ending
    :return: sorted array
    """
    if l < r:
        middle = (l + (r - 1)) // 2
        merge_sort(array, l, middle)
        merge_sort(array, middle+1, r)
        _merge(array, middle, l, r)
    return array

File: c++/sorting/merge_sort.cpp
//
//  insertion_sort.cpp
//  algorithms
//
//  Created by alifar on 8/8/16.
//  Copyright © 2016 alifar. All rights reserved.
//

#include "merge_sort.hpp"

void merge(std::vector<int> &arr, int left, int middle, int right){
	int i, j, k;
	int n1 = middle - left + 1;
	int n2 = right - middle;

	std::vector<int> left_part(n1);
	std::vector<int> right_part(n2);

	for(i = 0; i < n1; ++i){
		left_part[i] = arr[left + i];
	}
	for(j = 0; j < n2; ++j){
		right_part[j] = arr[middle + 1 + j];
	}

	i = 0;
	j = 0;
	k = left;

	while(i < n1 && j < n2){
		if(left_part[i] < right_part[j]){
			arr[k] = left_part[i];
			++i;
		} else{
			arr[k] = right_part[j];
			++j;
		}
		++k;
	}

	while(i < n1){
		arr[k] = left_part[i];
		++i;
		++k;
	}

	while(j < n2){
		arr[k] = right_part[j];
		++j;
		++k;
	}
}

void merge_sort_helper(std::vector<int> &array, int left, int right){
	if(left < right){
		int middle = left + (right - left)/2;

		merge_sort_helper(array, left, middle);
		merge_sort_helper(array, middle + 1, right);

		merge(array, left, middle, right);
	}
}

void merge_sort(std::vector<int> &array){
	merge_sort_helper(array, 0, array.size()-1);
}